message/html/markdown.js

```diff
@@ -15,29 +15,37 @@ exports.create = function (api) {
 
   function markdown (content) {
     if (typeof content === 'string') { content = {text: content} }
-    // handle patchwork style mentions.
+    // handle patchwork style mentions and custom emoji.
     var mentions = {}
+    var emojiMentions = {}
     if (Array.isArray(content.mentions)) {
       content.mentions.forEach(function (link) {
-        if (link.name) mentions['@' + link.name] = link.link
+        if (link && link.name && link.link) {
+          if (link.emoji) emojiMentions[link.name] = link.link
+          else mentions['@' + link.name] = link.link
+        }
       })
     }
 
     var md = h('div', {className: 'Markdown'})
     md.innerHTML = renderer.block(content.text, {
-      emoji: renderEmoji,
+      emoji: (emoji) => {
+        var url = emojiMentions[emoji]
+          ? api.blob.sync.url(emojiMentions[emoji])
+          : api.emoji.sync.url(emoji)
+        return renderEmoji(emoji, url)
+      },
       toUrl: (id) => {
         if (ref.isBlob(id)) return api.blob.sync.url(id)
-        return (mentions[id] ? mentions[id] : id)
+        return mentions[id] || id
       },
       imageLink: (id) => id
     })
 
     return md
   }
 
-  function renderEmoji (emoji) {
-    var url = api.emoji.sync.url(emoji)
+  function renderEmoji (emoji, url) {
```

Comment:
Need to use an escapeAttribute rather than escape url here, and then everything would be fine. Blob URL already handles the escaping. The reason we never noticed this before is that there was nothing to double encode in the old emoji urls since they were just file names.